### DISCUSSION TITLE
Notification client stop connecting after OTA upload

### DISCUSSION THREAD

**Discussion by meccip** [JAndrassy/ArduinoOTA#208] [General]
Hello,
I have mega2560 in use with some sensors, relay and light control. I use mqtt PubSubClient for sending data in/out on port 1883 and Wirepusher for notifications on port 80. If i upload OTA sketch into mega, my Wirepusher fail to connect on port 80 to send notification to my phone. When i remove OTA the connection is ok. Mqtt client is working with OTA.

This are some code line from the begining of the sketch:
byte mac[] = {0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA };
IPAddress ip(192, 168, 1, 112);                  // IP address,
IPAddress server(192, 168, 1, 117);             // mqtt server
EthernetServer androidServer(2562);              // set port number for android netIO
EthernetClient clientWirepusher;                           // client for notification
EthernetClient ethClient;
PubSubClient client(ethClient);

I think is a conflict between clients. So i try to start ArduinoOTA only when i need it by sending command on mqtt.
I have a ota state initialized false at arduino startup and when i need to upload OTA i send a true state on mqtt.
So, I have this:

#include <ArduinoOTA.h>
boolean otastart = false;
void runota() {
           ArduinoOTA.begin(Ethernet.localIP(), "arduino", "password", InternalStorage);
           otastart=true;
           }
void loop(){ 
 if (otastart==true){
  client.publish("ArduinoSlave/OTA","ready");
  ArduinoOTA.poll();
  delay(2000);
  return;
}
and from mqtt command (payload, "ota")  to activate runota();
}

But unfortunately it doesn't work all the time. Sometimes the program freezes when I start the ota function.
Any ideas are welcome.
Thank you.

**Comment by JAndrassy** [comment]
do you really want that `return` from `loop()` in `if (otastart==true){`?

and why do you wait 2 seconds every loop after `ArduinoOTA.poll();`?

**Reply by meccip** [reply]
That return is working only when i activate ota  to stop the program from running all code. After ota upload (when is working and not freeze) the return will not be active and program run normally.
I try without return and notification client start to send and freeze artduinoOta.
I have a delay of 5 sec when arduino start and send first notification. If i activate ota in the first 5sec is always working. After that 5sec when first notification is send, if i activate ota sometimes is freezing.

I wait 2sec thinking arduino is not freezing.

**Reply by meccip** [reply]
does arduinoOta have someting with port 80, because wirepusher is working with port 80 ?

**Reply by JAndrassy** [reply]
> does arduinoOta have someting with port 80, because wirepusher is working with port 80 ?

no. there is nothing with port 80. and you access a port on other device so there is no relation to local port 80

**Reply by meccip** [reply]
ok. But why is not working wirepusher client when arduinoOTA is running?
There is a limited number of clients that can run on Ethernet ?
Or is it possible when I activate OTA from mqtt to close wirepusher client or reinitialize it ?

**Comment by JAndrassy** [comment]
which Ethernet library is it? the Arduino Ethernet library for W5500?

https://github.com/arduino-libraries/Ethernet/blob/d31148d983ba61ac6340349ada8a0bfd09bd3cc0/src/Ethernet.h#L36-L40

**Reply by meccip** [reply]
I look into user/....arduino/libraries and Ethernet library looks like this. I have W5100.
I just remove mqtt PubSubClient from code, added arduinoOTA to run in loop and is working normally. My wirepusher is connecting and send notification.
So my conflict is with PubSub but what is strange that mqtt is working with ota and ota is blocking wirepusher.

**Reply by meccip** [reply]
ok. I think i have a way to make it work. I added in `void runota()` 2 lines that stop the mqtt client and wirepusher before ArduinoOta begin.

void runota() {
           client.publish("ArduinoSlave/OTA","ready");  //send mqtt that ota is ready
           client.disconnect(); // stop mqtt client
           clientWirepusher.stop(); // stop wirepusher client
           delay(2000);
           ArduinoOTA.begin(Ethernet.localIP(), "arduino", "password", InternalStorage);
           otastart=true;
                   } 

So far is not freezing. I will keep test it.
